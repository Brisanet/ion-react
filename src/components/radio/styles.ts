import { css, styled } from 'styled-components';

export const RadioGroup = styled.fieldset`
  border: none;
  padding: 0;
  margin: 0;
  display: flex;
  flex-direction: column;
  gap: 16px;
`;

export const Container = styled.div<{ $disabled?: boolean }>`
  ${({ theme, $disabled }) => css`
    ${theme.utils.flex.start(8)}

    label {
      ${theme.font.size[14]}
      font-weight: 400;
      color: ${theme.colors.neutral[7]};
      cursor: pointer;

      ${$disabled &&
      css`
        color: ${theme.colors.neutral[5]};
        cursor: not-allowed;
      `}
    }

    input[type='radio'] {
      appearance: none;
      width: 16px;
      height: 16px;
      margin: 4px;
      border-radius: 50%;
      border: 1px solid ${theme.colors.neutral[5]};
      background-color: ${theme.colors.neutral[1]};
      ${theme.utils.focus}
      cursor: pointer;

      &::before {
        content: '';
        display: none;
        width: 4px;
        height: 4px;
        border-radius: 50%;
        background-color: ${theme.colors.primary[2]};
        position: relative;
        top: 50%;
        left: 50%;
        transform: scale(1) translate(-50%, -50%);
      }

      &:hover,
      &:focus-visible {
        border: 1px solid ${theme.colors.primary[3]};
        background-color: ${theme.colors.primary[1]};
        box-shadow: 0 0 0 4px ${theme.colors.primary[1]};
      }

      &:hover::before {
        display: block;
      }

      &:active {
        border: 1px solid ${theme.colors.primary[5]};
        background-color: ${theme.colors.primary[2]};
        box-shadow: none;

        &::before {
          display: block;
          background-color: ${theme.colors.primary[1]};
        }
      }

      &:disabled {
        border: 1px solid ${theme.colors.neutral[4]};
        background-color: ${theme.colors.neutral[3]};
        cursor: not-allowed;
        box-shadow: none;

        &::before {
          display: none;
        }
      }

      &:checked {
        border: 4px solid ${theme.colors.main.primary};
        background-color: ${theme.colors.primary[1]};

        &:hover,
        &:focus-visible {
          border: 4px solid ${theme.colors.primary[5]};
          box-shadow: 0 0 0 4px ${theme.colors.primary[1]};

          &::before {
            display: none;
          }
        }

        &:active {
          border: 5px solid ${theme.colors.primary[7]};
          box-shadow: none;
        }

        &:disabled {
          border: 4px solid ${theme.colors.neutral[2]};
          background-color: ${theme.colors.neutral[4]};
          box-shadow: none;
        }
      }
    }
  `}
`;
