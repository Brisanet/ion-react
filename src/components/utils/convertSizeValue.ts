/**
 * Get a size value based on its type.
 * @param {string | number} value - The size value.
 * @returns The value itself if it's a string, or the value in pixels if it is a number.
 **/
export const convertSizeValue = (value: string | number): string =>
  typeof value === 'number' ? `${value}px` : value;
