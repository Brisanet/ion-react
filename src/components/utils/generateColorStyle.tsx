// Tipo para objeto de estilos  da função geradora de estilos
type ColorStyle = {
    background: string;
    color: string;
    border?: string;
  };
  
  /* 
    Função que gera cores de Background, Border e cor de
    fonte a fim de reduzir repetições de codigo
  
    Seguir o padrão de parâmetros: 
    1- Cor do background;
    2- Cor da fonte do botão;
    3- (Opicional) Cor da borda;
  */
 export function generateColorStyle(
    bgColor: string,
    fontColor: string,
    borderColor: string
  ) {
    let styles: ColorStyle = {
      background: bgColor,
      color: fontColor,
    };
  
    borderColor && (styles.border = `1px solid ${borderColor}`);
  
    return styles;
  }