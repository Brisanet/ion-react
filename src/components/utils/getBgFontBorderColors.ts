interface ObjStyles {
  background: string;
  color: string;
  border?: string;
}

/**
 * Function to prevent repetition of style blocks.
 *
 * @param background - Background color
 * @param color - The font color
 * @param border - Border attribute styles - example: ´1px solid blue´
 * @returns A Javascript object containing color properties.
 */
export function getBgFontBorderColors({
  background,
  color,
  border = '',
}: ObjStyles): ObjStyles {
  return {
    background,
    color,
    border,
  };
}
