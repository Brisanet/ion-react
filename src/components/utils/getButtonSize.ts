/**
 * Function to prevent repetition of style blocks.
 *
 * @param buttonSize - Height and min-width
 * @param padding - Button padding
 * @param fontSize - The size of font
 * @returns An Javascript object containing size properties.
 */
export function getButtonSize(
  buttonSize: string,
  padding: string,
  fontSize: string
) {
  let size = {
    height: buttonSize,
    minWidth: buttonSize,
    padding: padding,
    fontSize: fontSize,
  };

  return size;
}
