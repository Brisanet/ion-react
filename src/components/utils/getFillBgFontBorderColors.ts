type ColorStyleResponse = {
  border: string;
  backgroundColor: string;
  color: string;
  svg: {
    fill: string;
  };
};

interface StyleProps {
  svgFillColor: string;
  backgroundColor: string;
  fontColor: string;
  borderColor: string;
}

/**
 * Function to prevent repetition of style blocks.
 *
 * @param svgFillColor - Fill color to be used in SVG
 * @param backgroundColor - Background color
 * @param fontColor - The font color
 * @param borderColor - Border color
 * @returns A Javascript object containing style properties.
 */
export function getFillBgFontBorderColors({
  backgroundColor,
  fontColor,
  borderColor,
  svgFillColor,
}: StyleProps) {
  let styles: ColorStyleResponse = {
    backgroundColor: backgroundColor,
    color: fontColor,
    border: `1px solid ${borderColor}`,
    svg: {
      fill: svgFillColor,
    },
  };

  return styles;
}
