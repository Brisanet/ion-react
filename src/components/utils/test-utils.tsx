import theme from '@ion/styles/theme';
import { render, RenderResult } from '@testing-library/react';
import { ThemeProvider } from 'styled-components';

export const renderWithTheme = (children: React.ReactNode): RenderResult =>
  render(<ThemeProvider theme={theme}>{children}</ThemeProvider>);
