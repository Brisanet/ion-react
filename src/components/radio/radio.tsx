import { Container, RadioGroup } from './styles';

interface Option {
  label: string;
  value: string;
  disabled?: boolean;
}

export interface IonRadioProps {
  options: Option[];
  name: string;
  onChange: (value: string) => void;
  value?: Option['value'];
}

export const IonRadio = ({ name, options, value, onChange }: IonRadioProps) => {
  const handleChange = (value: string) => {
    onChange(value);
  };

  return (
    <RadioGroup>
      {options.map(({ label, value: optionValue, disabled }) => (
        <Container key={optionValue} $disabled={disabled}>
          <input
            id={optionValue}
            name={name}
            type='radio'
            value={optionValue}
            checked={optionValue === value}
            disabled={disabled}
            onChange={() => handleChange(optionValue)}
          />
          <label htmlFor={optionValue}>{label}</label>
        </Container>
      ))}
    </RadioGroup>
  );
};
