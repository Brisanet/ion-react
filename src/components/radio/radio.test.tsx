import userEvent from '@testing-library/user-event';
import { renderWithTheme } from '../utils/test-utils';
import { IonRadio, IonRadioProps } from './radio';

const options = [
  { label: 'Option 1', value: 'option1' },
  { label: 'Option 2', value: 'option2' },
];

const sut = (props: IonRadioProps) => {
  return renderWithTheme(<IonRadio {...props} />);
};

describe('IonRadio', () => {
  it('should render a given option selected', () => {
    const { container, getByRole } = sut({
      name: 'radio',
      options,
      value: 'option1',
      onChange: jest.fn(),
    });

    expect(getByRole('radio', { name: 'Option 1' })).toBeChecked();
    expect(container).toMatchSnapshot();
  });

  it('should render correctly when disabled', () => {
    const { container, getByRole } = sut({
      name: 'radio',
      options: [
        ...options,
        { label: 'Option 3', value: 'option3', disabled: true },
      ],
      value: 'option1',
      onChange: jest.fn(),
    });

    expect(getByRole('radio', { name: 'Option 3' })).toBeDisabled();
    expect(container).toMatchSnapshot();
  });

  it('should call onChange when an option is clicked', async () => {
    const onChange = jest.fn();

    const { getByLabelText } = sut({
      name: 'radio',
      options,
      value: 'option1',
      onChange,
    });

    await userEvent.click(getByLabelText('Option 2'));

    expect(onChange).toHaveBeenCalledWith('option2');
  });
});
