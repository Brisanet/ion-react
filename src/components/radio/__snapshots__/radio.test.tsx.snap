// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IonRadio should render a given option selected 1`] = `
.c0 {
  border: none;
  padding: 0;
  margin: 0;
  display: flex;
  flex-direction: column;
  gap: 16px;
}

.c1 {
  display: flex;
  align-items: center;
  justify-content: flex-start;
  gap: 8px;
}

.c1 label {
  font-size: 1.4rem;
  line-height: 2rem;
  font-weight: 400;
  color: #505566;
  cursor: pointer;
}

.c1 input[type='radio'] {
  appearance: none;
  width: 16px;
  height: 16px;
  margin: 4px;
  border-radius: 50%;
  border: 1px solid #aeb2bd;
  background-color: #fcfcfd;
  cursor: pointer;
}

.c1 input[type='radio']:focus-visible {
  outline: 2px solid #146ff5;
  outline-offset: 2px;
}

.c1 input[type='radio']::before {
  content: '';
  display: none;
  width: 4px;
  height: 4px;
  border-radius: 50%;
  background-color: #b5d2fc;
  position: relative;
  top: 50%;
  left: 50%;
  transform: scale(1) translate(-50%, -50%);
}

.c1 input[type='radio']:hover,
.c1 input[type='radio']:focus-visible {
  border: 1px solid #84b4fa;
  background-color: #ebf3fe;
  box-shadow: 0 0 0 4px #ebf3fe;
}

.c1 input[type='radio']:hover::before {
  display: block;
}

.c1 input[type='radio']:active {
  border: 1px solid #146ff5;
  background-color: #b5d2fc;
  box-shadow: none;
}

.c1 input[type='radio']:active::before {
  display: block;
  background-color: #ebf3fe;
}

.c1 input[type='radio']:disabled {
  border: 1px solid #ced2db;
  background-color: #e4e6eb;
  cursor: not-allowed;
  box-shadow: none;
}

.c1 input[type='radio']:disabled::before {
  display: none;
}

.c1 input[type='radio']:checked {
  border: 4px solid #0858ce;
  background-color: #ebf3fe;
}

.c1 input[type='radio']:checked:hover,
.c1 input[type='radio']:checked:focus-visible {
  border: 4px solid #146ff5;
  box-shadow: 0 0 0 4px #ebf3fe;
}

.c1 input[type='radio']:checked:hover::before,
.c1 input[type='radio']:checked:focus-visible::before {
  display: none;
}

.c1 input[type='radio']:checked:active {
  border: 5px solid #06439d;
  box-shadow: none;
}

.c1 input[type='radio']:checked:disabled {
  border: 4px solid #f2f3f5;
  background-color: #ced2db;
  box-shadow: none;
}

<div>
  <fieldset
    class="c0"
  >
    <div
      class="c1"
    >
      <input
        checked=""
        id="option1"
        name="radio"
        type="radio"
        value="option1"
      />
      <label
        for="option1"
      >
        Option 1
      </label>
    </div>
    <div
      class="c1"
    >
      <input
        id="option2"
        name="radio"
        type="radio"
        value="option2"
      />
      <label
        for="option2"
      >
        Option 2
      </label>
    </div>
  </fieldset>
</div>
`;

exports[`IonRadio should render correctly when disabled 1`] = `
.c0 {
  border: none;
  padding: 0;
  margin: 0;
  display: flex;
  flex-direction: column;
  gap: 16px;
}

.c1 {
  display: flex;
  align-items: center;
  justify-content: flex-start;
  gap: 8px;
}

.c1 label {
  font-size: 1.4rem;
  line-height: 2rem;
  font-weight: 400;
  color: #505566;
  cursor: pointer;
}

.c1 input[type='radio'] {
  appearance: none;
  width: 16px;
  height: 16px;
  margin: 4px;
  border-radius: 50%;
  border: 1px solid #aeb2bd;
  background-color: #fcfcfd;
  cursor: pointer;
}

.c1 input[type='radio']:focus-visible {
  outline: 2px solid #146ff5;
  outline-offset: 2px;
}

.c1 input[type='radio']::before {
  content: '';
  display: none;
  width: 4px;
  height: 4px;
  border-radius: 50%;
  background-color: #b5d2fc;
  position: relative;
  top: 50%;
  left: 50%;
  transform: scale(1) translate(-50%, -50%);
}

.c1 input[type='radio']:hover,
.c1 input[type='radio']:focus-visible {
  border: 1px solid #84b4fa;
  background-color: #ebf3fe;
  box-shadow: 0 0 0 4px #ebf3fe;
}

.c1 input[type='radio']:hover::before {
  display: block;
}

.c1 input[type='radio']:active {
  border: 1px solid #146ff5;
  background-color: #b5d2fc;
  box-shadow: none;
}

.c1 input[type='radio']:active::before {
  display: block;
  background-color: #ebf3fe;
}

.c1 input[type='radio']:disabled {
  border: 1px solid #ced2db;
  background-color: #e4e6eb;
  cursor: not-allowed;
  box-shadow: none;
}

.c1 input[type='radio']:disabled::before {
  display: none;
}

.c1 input[type='radio']:checked {
  border: 4px solid #0858ce;
  background-color: #ebf3fe;
}

.c1 input[type='radio']:checked:hover,
.c1 input[type='radio']:checked:focus-visible {
  border: 4px solid #146ff5;
  box-shadow: 0 0 0 4px #ebf3fe;
}

.c1 input[type='radio']:checked:hover::before,
.c1 input[type='radio']:checked:focus-visible::before {
  display: none;
}

.c1 input[type='radio']:checked:active {
  border: 5px solid #06439d;
  box-shadow: none;
}

.c1 input[type='radio']:checked:disabled {
  border: 4px solid #f2f3f5;
  background-color: #ced2db;
  box-shadow: none;
}

.c2 {
  display: flex;
  align-items: center;
  justify-content: flex-start;
  gap: 8px;
}

.c2 label {
  font-size: 1.4rem;
  line-height: 2rem;
  font-weight: 400;
  color: #505566;
  cursor: pointer;
  color: #aeb2bd;
  cursor: not-allowed;
}

.c2 input[type='radio'] {
  appearance: none;
  width: 16px;
  height: 16px;
  margin: 4px;
  border-radius: 50%;
  border: 1px solid #aeb2bd;
  background-color: #fcfcfd;
  cursor: pointer;
}

.c2 input[type='radio']:focus-visible {
  outline: 2px solid #146ff5;
  outline-offset: 2px;
}

.c2 input[type='radio']::before {
  content: '';
  display: none;
  width: 4px;
  height: 4px;
  border-radius: 50%;
  background-color: #b5d2fc;
  position: relative;
  top: 50%;
  left: 50%;
  transform: scale(1) translate(-50%, -50%);
}

.c2 input[type='radio']:hover,
.c2 input[type='radio']:focus-visible {
  border: 1px solid #84b4fa;
  background-color: #ebf3fe;
  box-shadow: 0 0 0 4px #ebf3fe;
}

.c2 input[type='radio']:hover::before {
  display: block;
}

.c2 input[type='radio']:active {
  border: 1px solid #146ff5;
  background-color: #b5d2fc;
  box-shadow: none;
}

.c2 input[type='radio']:active::before {
  display: block;
  background-color: #ebf3fe;
}

.c2 input[type='radio']:disabled {
  border: 1px solid #ced2db;
  background-color: #e4e6eb;
  cursor: not-allowed;
  box-shadow: none;
}

.c2 input[type='radio']:disabled::before {
  display: none;
}

.c2 input[type='radio']:checked {
  border: 4px solid #0858ce;
  background-color: #ebf3fe;
}

.c2 input[type='radio']:checked:hover,
.c2 input[type='radio']:checked:focus-visible {
  border: 4px solid #146ff5;
  box-shadow: 0 0 0 4px #ebf3fe;
}

.c2 input[type='radio']:checked:hover::before,
.c2 input[type='radio']:checked:focus-visible::before {
  display: none;
}

.c2 input[type='radio']:checked:active {
  border: 5px solid #06439d;
  box-shadow: none;
}

.c2 input[type='radio']:checked:disabled {
  border: 4px solid #f2f3f5;
  background-color: #ced2db;
  box-shadow: none;
}

<div>
  <fieldset
    class="c0"
  >
    <div
      class="c1"
    >
      <input
        checked=""
        id="option1"
        name="radio"
        type="radio"
        value="option1"
      />
      <label
        for="option1"
      >
        Option 1
      </label>
    </div>
    <div
      class="c1"
    >
      <input
        id="option2"
        name="radio"
        type="radio"
        value="option2"
      />
      <label
        for="option2"
      >
        Option 2
      </label>
    </div>
    <div
      class="c2"
    >
      <input
        disabled=""
        id="option3"
        name="radio"
        type="radio"
        value="option3"
      />
      <label
        for="option3"
      >
        Option 3
      </label>
    </div>
  </fieldset>
</div>
`;
